runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -866,80 +831,155 @@ TR_J9InlinerPolicy::createAnchorNodesForUnsafeGetPut(TR::TreeTop* treeTop,
                                        TR::TreeTop* callNodeTreeTop,
                                        TR::TreeTop* prevTreeTop,
                                        TR::SymbolReference* newSymbolReferenceForAddress,
                                        TR::TreeTop* directAccessTreeTop,
-                                       TR::TreeTop* lowTagCmpTree,
+                                       TR::TreeTop* arrayDirectAccessTreeTop,
+                                       TR::TreeTop* indirectAccessTreeTop,
                                        bool needNullCheck, bool isUnsafeGet,
                                        bool conversionNeeded,
-                                       TR::Block * joinBlock,
-                                       TR_OpaqueClassBlock *javaLangClass,
                                        TR::Node* orderedCallNode = NULL)
    {
    TR::CFG *cfg = comp()->getFlowGraph();
-   TR::Block *nullComparisonBlock = prevTreeTop->getEnclosingBlock();
-   TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
-   TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
-   TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
+
+   // Generate the tree for the null comparison
+   TR::Node *addrLoad =
+      TR::Node::createWithSymRef(unsafeAddress,
+                      comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()),
+                      0, newSymbolReferenceForAddress);
+   TR::Node *nullComparisonNode =
+      TR::Node::createif(TR::ifacmpeq, addrLoad,
+                        TR::Node::create(addrLoad, TR::aconst, 0, 0), NULL);
+   TR::TreeTop *nullComparisonTree = TR::TreeTop::create(comp(), nullComparisonNode);
+
+   // If low-order bit of offset is set, perform indirect access.  The IL for the case
+   // that handles conversion places the indirect access code on the fall-through
+   // path, so it needs to branch if the bit is not set; otherise, the indirect
+   // access code is on the taken path of the branch, so it needs to branch if the
+   // bit is set.
+   //
+   TR::TreeTop *lowTagCmpTree = genClassCheckForUnsafeGetPut(unsafeOffset, /* branchIfLowTagged */ !conversionNeeded);
+
+   TR::TreeTop *firstComparisonTree;
+   TR::TreeTop *ifTree;
+   TR::TreeTop *elseTree;
+
+   TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
+
+   // Determine overall layout of IL - which test to place first (null test of object or
+   // setting of low-order bit of offset) - and whether direct or indirect access IL will
+   // be placed on the fall-through path of the first branch.
+   //
+   if (conversionNeeded)
```

Comment:
Is the point of this if/else to keep the behavior unchanged for the "no conversion" cases ? i.e. < 32 bits ? If so, presumably this could also be relaxed in the future (not in this PR) since it is only a point in time limitation

Reply:
Right - it's keeping behaviour unchanged for the case where conversion might be needed as well as for the case where the J9Class for `java/lang/Class` is not known.

A future enhancement could arrange things for those two cases so that the `directAccessTreeTop` is on the fall-through path, and the first test performed is for the low-tagged offset bit.  That would also entail adjustments to the code that handles the placement of the test of whether the object is an array [below](https://github.com/eclipse-openj9/openj9/blob/c79d879a041f08d10b6ade76035c7fb4b8134122/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L971-L1012).